rolling-shutter/medley/broker.go

```diff
@@ -0,0 +1,184 @@
+package medley
+
+import (
+	"sync"
+
+	"github.com/rs/zerolog/log"
+)
+
+// `Broker` allows to distribute a value to multiple
+// receiving channels, so that all channels receive the same value.
+// This is different to the Fan-Out pattern where multiple channels
+// are listening for a send on one channel, but only the first channel
+// to select the receive operation receives this value.
+//
+// The Broker allows to set whether the internal send operation is
+// blocking or non blocking. In the blocking case, all receiving
+// (subscribed) channels have to successively select the receive in order
+// for the broker to continue operation. So if one receiver blocks,
+// everything else will block as well.
+// In the non-blocking case, a subscribed channel that is blocking
+// while the published value is sent will miss the value.
+//
+// One has to be careful in when to select which case.
+// The blocking behavior can be partially subverted by increasing
+// the subscribers buffer size - but if a subscriber is blocked forever,
+// this will only work until this subscribers buffer is full.
+type Broker[T any] struct {
+	mux                  sync.RWMutex
+	stop                 chan struct{}
+	publish              chan T
+	subscribe            chan chan T
+	unsubscribe          chan chan T
+	running              bool
+	bufferSizeSubscriber int
+	blockingSend         bool
+}
+
+func NewBroker[T any](buffer int, blocking bool) *Broker[T] {
+	return &Broker[T]{
+		stop:                 make(chan struct{}),
+		publish:              make(chan T, 1),
+		subscribe:            make(chan chan T),
+		unsubscribe:          make(chan chan T),
+		bufferSizeSubscriber: buffer,
+		blockingSend:         blocking,
+		running:              false,
+	}
+}
+
+// `Start` starts the Broker's internal loop that processes
+// the subscription and publish operations.
+// `initialSubscriptions` will be registered
+// before the loop is started.
+// This is useful when the channels must not miss a published value
+// after the start or when the subscribers buffer size
+// should diverge from the default value `bufferSizeSubscriber`
+// set in the constructor.
+func (b *Broker[T]) Start(initialSubscriptions ...chan T) {
```

Comment:
`initialSubscriptions` might have a different capacity than channels created with `Subscribe` which could lead to unexpected behavior. How about

- changing `Subscribe` to take a channel as a parameter (and thus get rid of `bufferSizeSubscriber`) or
- in `Start` take a parameter `numInitialSubscriptions` and creating and returning the channels with the expected size

?

Reply:
The reason for the initialSubscriptions is the following:

I wanted to be able to have a receive on the subscribed channels waiting in a (or multiple) goroutine and then start the broker. That way it is impossible to miss the first published value.
This would not be easily possible with your second suggestion.

Another way to achieve this would be to make the Subscribe( ) append to the initial subscriptions directly when the broker is stopped, which comes close to your suggestion.
It's possible, but then we would have a `subscriptions` field on the object rather than local state in the Start( ) function.

Currently I favour your first suggestion because this would remove the inconsistency and the slightly confusing default buffer-size. It would also provide more flexibility to the caller of `Subscribe( )`

Reply:
Now the `Subscribe()` takes a channel argument, the buffer size is removed, and subscribe appends to a list of initial subscriptions when the loop is not running.